core/src/test/scala/org/apache/spark/sql/txn/TxnTestSuite.scala

```diff
@@ -0,0 +1,234 @@
+  )
+  protected final val giveString = resourceToString(
+    s"resolveLock-test/1_give.sql",
+    classLoader = Thread.currentThread().getContextClassLoader
+  )
+  protected final val getString = resourceToString(
+    s"resolveLock-test/2_get.sql",
+    classLoader = Thread.currentThread().getContextClassLoader
+  )
+  protected final val accountString = resourceToString(
+    s"resolveLock-test/1_account.sql",
+    classLoader = Thread.currentThread().getContextClassLoader
+  )
+  protected final val rnd = new scala.util.Random
+
+  /**
+   * query to tidb with jdbc and txn style
+   *
+   * @param query all querys run in the txn
+   * @param wait whether wait every 2 seconds
+   * @return Unit
+   */
+  protected def queryTIDBTxn(query: Seq[String], wait: Boolean): Unit = {
+    val conn = DriverManager.getConnection(jdbcUrl, "root", "")
+    try {
+      //Assume a valid connection object conn
+      conn.setAutoCommit(false)
+      val stmt = conn.createStatement()
+      query.foreach {
+        case q: String =>
+          stmt.executeUpdate(q)
+          if (wait)
+            Thread.sleep(2000)
+      }
+      conn.commit()
+    } catch {
+      case e: SQLException =>
+        logger.info("rollback1" + e.getMessage)
+        conn.rollback()
+        throw e
+    }
+  }
+
+  /**
+   * get Spark query thread using q1 query
+   *
+   * @param i the query number
+   * @return thread
+   */
+  protected def firstQueryThread(i: Int): Thread =
+    new Thread {
+      override def run {
+        var ok = true
+        while (ok) {
+          try {
+            querySpark(q1String)
+            logger.info("query1 " + i.toString + " success!")
+            ok = false
+          } catch {
+            case _: SQLException =>
+              Thread.sleep(1000 + rnd.nextInt(3000))
+              ok = true
+          }
+        }
+      }
+    }
+
+  /**
+   * get Tidb txn thread using 2 transactions(A => B, A - x and B + x)
+   *
+   * @param i the query number
+   * @return thread
+   */
+  protected def firstTxnThread(i: Int): Thread =
+    new Thread {
+      override def run {
+        var ok = true
+        while (ok) {
+          try {
+            val num = rnd.nextInt(600).toString
+            val id1 = (1 + rnd.nextInt(150)).toString
+            val id2 = (1 + rnd.nextInt(150)).toString
+            val querys = Seq[String](
+              giveString.replace("$1", num).replace("$2", id1),
+              getString.replace("$1", num).replace("$2", id2)
+            )
+            queryTIDBTxn(querys, true)
+            logger.info("txn1 " + i.toString + " success!")
+            ok = false
+          } catch {
+            case _: SQLException =>
+              Thread.sleep(1000 + rnd.nextInt(3000))
+          }
+        }
+      }
+    }
+
+  /**
+   * get Spark query thread using q2 query, a long running transaction
+   *
+   * @param i the query number
+   * @return thread
+   */
+  protected def secondQueryThread(i: Int): Thread =
```

Comment:
It is the same as `firstQueryThread` except for its query string, isn't it? You should merge them to avoid duplicate code